sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,376 @@
+     */
+    Builder include(HasDisplayData subComponent);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, String value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, long value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, double value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Instant value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Duration value);
+
+    /**
+     * Register the given display metadata.
+     */
+    ItemBuilder add(String key, Class<?> value);
+  }
+
+  /**
+   * Utility to append optional fields to display metadata, or register additional display metadata
+   * items.
+   */
+  public static interface ItemBuilder extends Builder {
+    /**
+     * Add a human-readable label to describe the metadata field. A label is optional; if
+     * unspecified, UIs should display the metadata key to identify the display item.
+     */
+    ItemBuilder withLabel(String label);
+
+    /**
+     * Add a link URL to the registered display metadata. A link URL is optional and
+     * can be provided to point the reader to additional details about the metadata.
+     */
+    ItemBuilder withLinkUrl(String url);
+  }
+
+  /**
+   * A display metadata item. DisplayData items are registered via {@link Builder#add} within
+   * {@link HasDisplayData#populateDisplayData} implementations. Each metadata item is uniquely
+   * identified by the specified key and namespace generated from the registering component's
+   * class name.
+   */
+  public static class Item<T> {
+    private final String key;
+    private final String ns;
+    private final Type type;
+    private final T value;
+    private final String label;
+    private final String url;
+
+    private static <T> Item<T> create(String namespace, String key, Type type, T value) {
+      return new Item<>(namespace, key, type, value, null, null);
+    }
+
+    private Item(String namespace, String key, Type type, T value, String url, String label) {
+      this.ns = namespace;
+      this.key = key;
+      this.type = type;
+      this.value = value;
+      this.url = url;
+      this.label = label;
+    }
+
+    String getNamespace() {
+      return ns;
+    }
+
+    Type getType() {
+      return type;
+    }
+
+    T getValue() {
+      return value;
+    }
+
+    String getLabel() {
+      return label;
+    }
+
+    String getUrl() {
+      return url;
+    }
+
+    String getKey() {
+      return key;
+    }
+
+    /**
+     * Retrieve a serialized version of the display item value suitable for use as JSON.
+     */
+    String getValueString() {
+      return type.serializeJsonString(value);
+    }
+
+    private Item<T> addLabel(String label) {
```

Comment:
Since these return new instances, the typical name would be `withLabel`. `add` suggests mutation.